common/activecluster/manager.go

```diff
@@ -27,83 +27,198 @@ import (
 	"errors"
 	"fmt"
 	"sync"
+	"sync/atomic"
 	"time"
 
 	"github.com/uber/cadence/common/cache"
+	"github.com/uber/cadence/common/clock"
 	"github.com/uber/cadence/common/cluster"
+	"github.com/uber/cadence/common/constants"
 	"github.com/uber/cadence/common/log"
 	"github.com/uber/cadence/common/log/tag"
 	"github.com/uber/cadence/common/metrics"
 	"github.com/uber/cadence/common/types"
 )
 
+const (
+	// notifyChangeCallbacksInterval is the interval at which external entity change callbacks are notified to subscribers.
+	// This is to avoid sending too many notifications and overwhelming the subscribers (i.e. per-shard history engines).
+	notifyChangeCallbacksInterval = 5 * time.Second
+)
+
 type DomainIDToDomainFn func(id string) (*cache.DomainCacheEntry, error)
 
-type manager struct {
-	domainIDToDomainFn DomainIDToDomainFn
-	clusterMetadata    cluster.Metadata
-	metricsCl          metrics.Client
-	logger             log.Logger
-	ctx                context.Context
-	cancel             context.CancelFunc
+type managerImpl struct {
+	domainIDToDomainFn      DomainIDToDomainFn
+	clusterMetadata         cluster.Metadata
+	metricsCl               metrics.Client
+	logger                  log.Logger
+	ctx                     context.Context
+	cancel                  context.CancelFunc
+	wg                      sync.WaitGroup
+	externalEntityProviders map[string]ExternalEntityProvider
+	timeSrc                 clock.TimeSource
+
+	shouldNotifyChangeCallbacks int32
+	changeCallbacksLock         sync.Mutex
+	changeCallbacks             map[int]func(ChangeType)
+
+	// define some internal helper functions as member variables to be mocked in tests
+	getWorkflowActivenessMetadataFn func(ctx context.Context, domainID, wfID, rID string) (*WorkflowActivenessMetadata, error)
+}
 
-	// TODO: fakes to be remove
-	wf1StartTime  time.Time
-	wf1FailedOver int32
+type ManagerOption func(*managerImpl)
 
-	changeCallbacksLock sync.RWMutex
-	changeCallbacks     map[int]func(ChangeType)
+func WithTimeSource(timeSource clock.TimeSource) ManagerOption {
+	return func(m *managerImpl) {
+		if timeSource != nil {
+			m.timeSrc = timeSource
+		}
+	}
 }
 
 func NewManager(
 	domainIDToDomainFn DomainIDToDomainFn,
 	clusterMetadata cluster.Metadata,
 	metricsCl metrics.Client,
 	logger log.Logger,
-) Manager {
+	externalEntityProviders []ExternalEntityProvider,
+	opts ...ManagerOption,
```

Comment:
nit: We're migrating packages to use Fx Modules. The functional options pattern doesn't align with Fx Module.s

Reply:
I think you can provide a constructor to Fx and that's where we can pass the options. We can revisit when resource_impl.go is moving to Fx